Add suppport for Apple- / Google-Pay to the checkout

trm217 opened the pull request:
# Tasks

## Checkout

- [x] Cleanup code
- [x] Improve logic of the first paymet-request initialization
- [x] Add a loading status to usePaymentRequest-hook and don't render submit-button while status=loading.
- [x] Normalize address data received from Payment UI
- [x] Delay complete('sucess') call until payPledege has succeeded.
- [x] Define final style of the Apple Pay button
- [x] Add Google Pay (adding button + slight adaptations to functionalities that handle Apple Pay)
- [x] Add message if payment-method is temporarily unavailable

## Konto
- [x] ~Show Apple Pay icon next to the Credit-Card Icon in~ 
- [x] ~Ensure that Apple Pay is not shown as an option when adding adding a new payment-source~ (No changes necessary, Apple Pay is only available in the PaymentForm) 

## Backend
- [x] ~Update transactional-emails to mention when a wallet-card is used.~ (Since we don't especially differentiate between Apple Pay and card payments, this isn't really needed)

## Finishing touches
- [x] delete console.debug statements in Submit.js and usePaymentRequest.tsx

tpreusse commented on apps/www/pages/angebote.js:
```diff
@@ -89,4 +89,4 @@ class PledgePage extends Component {
 }
 
-export default withDefaultSSR(withRouter(PledgePage))
+export default withRouter(PledgePage)
```
please re-add

trm217 replied:
Alright, this will however cause Apple-Pay to not be initially available on the first-render.

tpreusse commented on apps/www/components/Payment/Form/ApplePay.tsx:
```diff
@@ -0,0 +1,39 @@
+// Check if the website is running in a secure document.
+function isInSecureDocument(): boolean {
+  return (
+    typeof window !== 'undefined' && window?.location?.protocol === 'https:'
```
```suggestion
    typeof window !== 'undefined' && window.location.protocol === 'https:'
```

tpreusse commented on apps/www/components/Payment/Form/ApplePay.tsx:
```diff
@@ -0,0 +1,39 @@
+import { useEffect, useState } from 'react'
+
+// Check if the website is running in a secure document.
+function isInSecureDocument(): boolean {
```
`isInSecureWindow` or `isInSecureClientWindow` would make it more obvious that this only return true on the client.

tpreusse commented on apps/www/components/Payment/Form/ApplePay.tsx:
```diff
@@ -0,0 +1,39 @@
+}
+
+function isApplePayAvailable(): boolean {
+  if (!window?.ApplePaySession) {
```
```suggestion
  if (!window.ApplePaySession) {
```

trm217 replied:
Wouldn't this cause an error when a Browser has not defined ApplePaySession?

tpreusse commented on apps/www/components/Payment/PaymentRequest/PaymentRequestEventHelper.ts:
```diff
@@ -0,0 +1,48 @@
+import {
+  PaymentMethod,
```
unused?

tpreusse commented on apps/www/components/Payment/PaymentRequest/PaymentRequestEventHelper.ts:
```diff
@@ -0,0 +1,48 @@
+export function getPayerInformationFromEvent(
+  event: PaymentRequestPaymentMethodEvent,
+): PayerInformation {
+  const [firstName, lastName] = event.payerName.split(' ').map((s) => s.trim())
+
+  return {
+    email: event.payerEmail,
+    firstName,
+    lastName,
```
Let's save everything. And I'd gamble that multiple first names are more common than space separated last names like `von Wunschheim`:

```suggestion
  const [...firstNames, lastName] = event.payerName.split(' ')

  return {
    email: event.payerEmail,
    firstName: firstNames.join(' ').trim(),
    lastName: lastName.trim(),
```

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -129,17 +132,60 @@ const SubmitWithHooks = (props) => {
+      amount: props.total ?? 0,
+      label: props.query.package ?? 'NO VALUE',
+    },
+    requestShipping: props?.requireShippingAddress ?? false,
+    shippingOptions: props?.requireShippingAddress
```
```suggestion
    requestShipping: props.requireShippingAddress || false,
    shippingOptions: props.requireShippingAddress
```

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -129,17 +132,60 @@ const SubmitWithHooks = (props) => {
+  const paymentRequest = useStripePaymentRequest({
+    total: {
+      amount: props.total ?? 0,
+      label: props.query.package ?? 'NO VALUE',
```
let's start with the package title
```
      label: t.first([`package/${props.query.package}/title`, 'package/choose']),
```

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -511,14 +634,15 @@ class Submit extends Component {
     } = this.state
 
-    if (paymentMethod !== 'STRIPE') {
+    if (paymentMethod !== 'STRIPE' && !paymentMethod?.startsWith('STRIPE-')) {
```
```suggestion
    if (!paymentMethod || !paymentMethod.startsWith('STRIPE')) {
```

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -532,11 +656,12 @@ class Submit extends Component {
       values: { paymentMethod },
     } = this.state
-    if (paymentMethod !== 'STRIPE') {
+    if (paymentMethod !== 'STRIPE' && !paymentMethod?.startsWith('STRIPE-')) {
```
```suggestion
    if (!paymentMethod || !paymentMethod.startsWith('STRIPE')) {
```

tpreusse reviewed (commented):
maybe we should call it `STRIPE-WALLET-APPLEPAY` and then do `STRIPE-WALLET` checks instead of `STRIPE-` which seem cryptic

is the `billing_details.address` visible client side or would that be something our backend has to grab?
<img width="951" alt="Screenshot 2022-02-22 at 22 40 11" src="https://user-images.githubusercontent.com/410211/155224354-09ab40c9-2a1a-4328-bb95-29dbd3bad6ed.png">

trm217 commented:
The billing address is not available in the [PaymentResponse object](https://stripe.com/docs/js/appendix/payment_response).

tpreusse commented:
I just tested ngrok and works on my machine too.

We need to move the «Schon mal ausgefüllt?» upwards, because after login one might also have a saved credit card, and rename it to some thing like «Haben Sie bereits ein Konto? Melden Sie sich an.»
<img width="1279" alt="Screenshot 2022-03-16 at 08 52 40" src="https://user-images.githubusercontent.com/410211/158542933-e0dc1d2f-ee74-445a-8822-bb6802a5d164.png">

When choosing pay by bill we should first ask for email and name before the full address:
<img width="1279" alt="Screenshot 2022-03-16 at 08 53 53" src="https://user-images.githubusercontent.com/410211/158543368-5e65473a-9b09-48a2-b6f1-11c46825bae8.png">

Maybe we could do that for all payment options? And even switch it around that we first ask for email and then derive first name and last name from the email to pre fill name fields? (when e.g. username of email is separated by a dot and contains at least two chars on each side)

trm217 commented:
Regarding the prefilling of the name based on the email: I wouldn't do that. Unless we have a bad implementation of the form fields, the prefilling would be suggested by every modern browser anyways. This would also cause issues with wrong prefills: for example fname and lname might be reversed in the email and thus the prefilling would be wrong, which then could easily be overseen by the user.

trm217 commented:
<img width="1089" alt="image" src="https://user-images.githubusercontent.com/30313631/158549859-3c6ebb50-3f01-44f2-b8a0-b137e9a4639d.png">
I updated the UI as requested, I also removed some whitespace.

tpreusse reviewed (commented):
I refactored the `selectedPaymentMethod` to be available in `SubmitWithHooks` – let's move this code to the `usePaymentRequest` hook ;)

Spacing when apple pay is selected is too small, add 10 to 30 more pixels between payment method & checkboxes:

<img width="685" alt="Screenshot 2022-03-17 at 13 24 49" src="https://user-images.githubusercontent.com/410211/158827604-93e81ef9-6366-4fe6-89b9-f831e8beae91.png">


Currently shipping address is above payment method selection: should be below (since hidden when apple pay)

<img width="1319" alt="Screenshot 2022-03-17 at 13 35 06" src="https://user-images.githubusercontent.com/410211/158827420-5336c55f-2c8b-4c0b-9d1e-651a3b6756e2.png">

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -610,151 +898,165 @@ class Submit extends Component {
+                this.props.selectedPaymentMethod &&
+                !this.isStripeWalletPayment() && (
+                  <div {...styles.topMargin}>
+                    {contactPreface && (
+                      <div style={{ marginBottom: 40 }}>
+                        <P>{contactPreface}</P>
+                      </div>
+                    )}
```
We should move this preface above the payment question:
<img width="795" alt="Screenshot 2022-03-18 at 11 08 19" src="https://user-images.githubusercontent.com/410211/158983593-189f1f85-24ad-4bee-9296-ff9e8a88ce52.png">

trm217 replied:
![Uploading image.png…]()
Done

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -172,6 +236,110 @@ class Submit extends Component {
+          this.setState(() => ({
+            loading: false,
+            walletError: t(
+              'account/pledges/payment/methods/initialization/error',
+            ),
+          }))
+        })
+
+      // Re-initialize the payment-request for the current wallet
+      this.setState(() => ({
+        loading: t('account/pledges/payment/methods/loading'), // The loading state is tracked though paymentRequest.status
+      }))
+    } else if (
+      isUnavailableAndStripeWalletIsSelected &&
+      !this.state.walletError
+    ) {
+      this.setState(() => ({
+        loading: false,
+        walletError: t('account/pledges/payment/methods/unavailable'),
+      }))
+    }
+    if (
+      this.props.selectedPaymentMethod &&
+      !this.props.selectedPaymentMethod.startsWith('STRIPE-WALLET') &&
+      this.state.walletError
+    ) {
+      this.setState(() => ({
+        walletError: null,
+      }))
+    }
+    if (
+      this.props.selectedPaymentMethod &&
+      !this.props.selectedPaymentMethod.startsWith('STRIPE') &&
+      this.state.stripeError
+    ) {
+      this.setState(() => ({
+        stripeError: null,
+      }))
+    }
+  }
```
I would have imagined that we move this whole logic to the `usePaymentRequest` hook

trm217 replied:
In case we do that, we would have to tear most of the Submit-component state apart and move it up to Submit Hooks. If we were to do that, we might as well refactor the entire Submit-component and convert it into a functional-component.

tpreusse replied:
are you sure? As far as I can see it's only `loading`, `walletError`, `stripeError` – afaik our `PaymentRequestValues` already have a loading state and could be extended with two error states?

trm217 replied:
Done

tpreusse commented on apps/www/components/Pledge/Submit.js:
```diff
@@ -172,6 +236,110 @@ class Submit extends Component {
+              walletError: null,
+            }))
+          } else {
+            const isGooglePayPayment =
+              this.props.selectedPaymentMethod ===
+              WalletPaymentMethod.GOOGLE_PAY
+
+            if (isGooglePayPayment) {
+              setSelectedPaymentMethod('STRIPE')
+            }
+            this.setState((prevState) => ({
+              ...prevState,
+              loading: false,
+              walletError: !isGooglePayPayment
+                ? t('account/pledges/payment/methods/unavailable')
+                : null,
+              stripeError: isGooglePayPayment
+                ? t('account/pledges/payment/methods/google-pay/unavailable')
+                : null,
+            }))
```
IMO this logic could be made generic, for the odd case where is happens with apple pay

trm217 replied:
Done, now works for both Apple and Google Pay (also works if other Wallets are added)

tpreusse reviewed (approved):
Let's ship this

trm217 commented on apps/www/components/Payment/Form/StripeWalletHelpers.tsx:
```diff
@@ -0,0 +1,64 @@
+
+  useEffect(() => {
+    // Only check for Google Pay on the client and when a secure connection is provided
+    setIsAvailable(isInSecureWindow() && isGooglePayAvailable())
```
Allright let's remove it.  Might cause confusion in dev though since Google Pay might show up in http.

tpreusse commented on apps/www/components/Payment/Form/StripeWalletHelpers.tsx:
```diff
@@ -0,0 +1,64 @@
+}
+
+function isApplePayAvailable(): boolean {
+  if (!window.ApplePaySession) {
```
don't forget to add it here ;-)
```suggestion
  if (!isInSecureWindow() && !window.ApplePaySession) {
```

trm217 replied:
Shouldn't be an issue since ApplePaySession isn't defined on non HTTPS. But "doppelt hält besser" xD
